### ISSUE TITLE
Atheros ath9k WiFi adapter

### ISSUE THREAD

**Issue by vipere** [01org/wysiwidi#86]
Just wanted to let you guys know the sink also works with my Atheros AR9287 based WiFi adapter, using the ath9k driver and my Nexus 4.

At first, there were some connection issues which I've worked around, and now, after I've updated to the latest wpa_supplicant (I noticed there were some interesting changes in the last couple of days) the workarounds are no longer necessary and it works out of the box.
You might want to update your wpa_supplicant version recommendation.

I do have some issues with gstreamer, but that's another matter, though if you could document which plugins are needed that would help. Currently I have no sound, and there are some decoding errors (libav stuff).

**Comment by kanavin** [comment]
What are you using as the source?

**Comment by vipere** [comment]
@kanavin A Nexus 4 running Android 4.4.4

**Comment by jku** [comment]
\o/  Great to hear, thanks for reporting.

I'll update the supplicant commit recommendation.

Gstreamer playback is certainly not perfect here either -- I've seen the libav errors as well, and I think they (or some of them) are a result of missing UDP packets although I've not gone through the trouble of actually confirming that 100%. Enabling RTSP*) in the GStreamer pipeline might help but we've not really had the time to get that tested. 

Sound works here though: I believe the pipeline should end up with GstAacParse and GstFaad for the decoding. @kanavin knows the GStreamer side better though...
 
*) If anyone intends to have a go at pipeline improvements, it makes sense to read Alexanders thoughts on an actual gstreamer element: https://lists.01.org/pipermail/wysiwidi-dev/2015-February/000007.html

**Comment by kanavin** [comment]
'RTSP' in the message above should be RTCP. Also, we're not using rtpjitterbuffer element at the moment on the sink side; there is no conclusive evidence that it helps much and you can't convince playbin to autoplug it (but it certainly wouldn't hurt).

Miracast is using h264 for video and lpcm, aac or ac3 for audio. If you have gst-libav installed, you should have decoders for everything. I'm not sure what Android is using for audio, but you can try running

mirac_network/gst-test --device=sink --port=5000
and then
gst-test --device=testsource --stream=both --hostname=127.0.0.1 --port=5000

If you get a picture with audio then at least ac3 should work.

**Comment by jku** [comment]
oops, fixed the typo. thanks

**Comment by jku** [comment]
So I tested this as well: indeed the my Sony USB WLAN adapter (ath9k_htc) now mostly works at least in a quick test.

Speaking of USB, I can see multiple bugs popping up because no-one's tested removing and inserting devices while P2P is happening... I've got a connman crash already and I bet our P2P::Client isn't robust about technologies appearing and disappearing either.

**Comment by vipere** [comment]
Thanks for the info @kanavin, @jku. I do seem to have the right plugins, but still no audio, even in the local test you suggested, so it's just a problem with my installation I suppose.

This led me to try VLC instead of gstreamer, and surprisingly it works just as well (and with sound).
Note that the decoding errors remain also with VLC.

I think jku's hunch is right - those errors are probably the result of network problems.
Pinging the source while it's streaming, shows there can be a high latency (up to a second even), but that doesn't seem to be the only problem, as I tried increasing the "Network buffer" in VLC (might be called jitter buffer elsewhere) and the errors remain.
I got the idea of disconnecting the source from the local WiFi network, leaving only the P2P network active. This has decreased the latency substantially, and eliminated many of the errors. Although playback is still not perfect, it's getting there.

I'm interested to know if you notice the same behavior on your side.
Another test could be using a different sink, perhaps they have ways to alleviate the network issues.

**Comment by kanavin** [comment]
Network buffer might help with packets that are out of order or late, but it won't help with packets that are fully lost. For that RTP has an optional facility called RTCP (which is a way to request a re-sending of missing packets). If VLC supports it you can try and see if it helps. In our local tests with android as the source the quality is just good enough - with occasional slight artefacts or stutter, that we didn't want to take the trouble of figuring out how to use RTCP with gstreamer.

**Comment by jku** [comment]
> I got the idea of disconnecting the source from the local WiFi network, leaving only the P2P
> network active. This has decreased the latency substantially, and eliminated many of the errors.

Yes, I can see both effects. The other major influence on picture quality might be chatter in the channel: If I test this in the office with dozens of people working around me I seem to get more artefacts while testing in the same spot after hours when no-one else is around might result in almost perfect playback.

Other sinks (commercial products) can play the same stream in these network conditions almost flawlessly (although their latency is in the same ball park) so we should be able to fix the playback problems. Like kanavin said, there's still things we can try on the gstreamer side but unfortunately changes in drivers or supplicant may be required ... as you've probably noticed they aren't exactly product ready with regards to P2P.

I've been meaning to do some kind of basic latency/bandwidth testing between two machines running connman+supplicant just to make sure we don't hit some basic bugs there: like you said sometimes ping times etc look a bit fishy... But so far there have been bigger fish to fry.

**Comment by masterajay** [comment]
Hi Everyone,
For a Wi-Fi Display to work , does the Wi-Fi Adaptor (Sink) should have any special  hardware feature apart from P2P support ?
I have a Wi-Fi USB adaptor  from TP-LINK (TL-WN722N ). Basic Wi-Fi P2P works with this , but when I try to set wifi_display=1 , it fails. This made me think only P2P may not be enough for Wi-Fi Display .
I use wpa_supplicant 2.5 with nl80211 driver interface.The USB adapotr uses ath9k_htc driver and htc_9271 firmware.

Can anyone suggest me a USB WiFI adaptor which I can use for WiFi dispaly development work under linux (kernel 3.16) ?

**Comment by hardiksingh-rathore** [comment]
Hi Vipere, Request to share the link for AR9287 USB  WiFi adpater to buy online.
